File: Olympiad/COCI/official/2007/regional_solutions/firefly.cpp

/*
  Croatian Regional Competition in Informatics 2007
  Task FIREFLY
  Programming language C++
*/

#include <algorithm>
#include <cstdio>

using namespace std;

int n, h;
int stalagmiti[100000];
int stalaktiti[100000];

int main( void ) {
   scanf( "%d%d", &n, &h ); n /= 2;
   for( int i = 0; i < n; i++ ) scanf( "%d%d", &stalagmiti[i], &stalaktiti[i] );

   sort( stalagmiti, stalagmiti+n );
   sort( stalaktiti, stalaktiti+n );

   int mini = 1000000000, broj = 0;

   for( int y = 0; y < h; ++y ) {
      int kungfu =
         n - (int)(lower_bound( stalagmiti, stalagmiti+n, y+1 )-stalagmiti) +
         n - (int)(lower_bound( stalaktiti, stalaktiti+n, h-y )-stalaktiti);

      if( kungfu  < mini ) { mini = kungfu; broj = 0; }
      if( kungfu == mini ) ++broj;
   }

   printf( "%d %d\n", mini, broj );

   return 0;
}

File: Olympiad/COCI/official/2007/regional_solutions/firefly.pas

(*
  Croatian Regional Competition in Informatics 2007
  Task FIREFLY
  Programming language Pascal
*)

program buba;

const MAXH = 500000;

var
   i, n, h, duljina                           : longint;
   najmanjePrepreka, brojNacina, brojPrepreka : longint;
   gore, dolje                                : array[1..MAXH] of longint;

begin

    readln(n, h);

    for i:=1 to h do
    begin
        gore[i]:=0;
        dolje[i]:=0;
    end;

    for i:=1 to n div 2 do
    begin
        readln(duljina);
        inc(dolje[duljina]);
        readln(duljina);
        inc(gore[duljina]);
    end;

    for i:=h-2 downto 1 do
    begin
        gore[i] := gore[i] + gore[i+1];
        dolje[i] := dolje[i] + dolje[i+1];
    end;

    najmanjePrepreka := n;
    brojNacina := 0;

    for i:=1 to h do
    begin
        brojPrepreka := dolje[i] + gore[h-i+1];
        if brojPrepreka < najmanjePrepreka then
        begin
            najmanjePrepreka := brojPrepreka;
            brojNacina := 1;
        end else
            if brojPrepreka = najmanjePrepreka then
                inc(brojNacina);
    end;

    writeln(najmanjePrepreka, ' ', brojNacina);

end.
